design-proposals/klip-31-metastore-backups.md

```diff
@@ -0,0 +1,122 @@
+* `ksql.backup.location` as String to specify a directory location for backups (default: an internal KSQL directory)
+
+## Design
+
+When backups are enabled, KSQL will start replaying the command_topic messages to a local file in the specified backup location. Every new command added to the
+command_topic while KSQL is running will be appended to the local file to keep a live metastore backup.
+
+The `ksql.enable.backup` config will be used to enable backups. It is better to let users enable/disable backups with this config than disabling by setting an empty backup location.
+Otherwise, a user setting the location empty just to disable a backup might forgot what the location was when enabling it back.
+
+The `ksql.backup.location` will specify the directory where the backup files exist. Only a directory in the local file system is accepted (no remote locations). This is to avoid multiple
+nodes in the KSQL cluster writing to the same remote location. The location is a directory where KSQL will write one or more files.
+
+The backup file name will be formed using the command_topic name and the date of the initial backup (i.e. _confluent-ksql-default__command_topic_1593540976.bak).
+
+The messages will be serialized using the command_topic JSON serializer. The backup file content will look pretty similar to what is seen in the command_topic messages. It will have one
+command per line. Each line will start with the same command_topic row timestamp to allow users identify the backup file with the most recent entries (in case of multiple nodes).
+
+i.e. (backup: _confluent-ksql-default__command_topic_1593545289.bak)
+```
+1593545289 {"statement":"CREATE STREAM TEST1(ID INT) WITH(KAFKA_TOPIC='test1', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545382 {"statement":"CREATE STREAM TEST2(ID INT) WITH(KAFKA_TOPIC='test2', VALUE_FORMAT='JSON');","streamsProperties":{}, ... }
+1593545529 {"statement":"DROP STREAM TEST1;","streamsProperties":{}, ... }
+```
+
+This format will allow users to easily restore the command_topic using the Kafka producer scripts.
+
+i.e.
+```
+$ cut -d' ' -f2- _confluent-ksql-default__command_topic_1593545289.bak | 
+      kafka-console-producer --broker-list localhost:9092 --topic _confluent-ksql-default__command_topic
+```
+
+### Workflow
+
+When KSQL starts, it always reads the command topic from the beginning. If the backup file does not exist yet, then it will create a new file and start appending each command_topic message to it.
+A backup file is not found if it is the first time backups are enabled, or if the KSQL service.id is different from the ones found in the backup filenames.
+
+If a backup file with the same KSQL service.id exists, then KSQL will append only the new command_topic messages to that file.
+
+KSQL will compare each consumed command_topic message against the next line from the backup file. If both messages are different, then it considers the command_topic as a new restored
```

Comment:
What's the comparison function? String equality? Compare the deserialized json objects on both sides? Something else?

Reply:
It will compare the deserialized objects. If we use Strigs, then we might get false equality if the command properties are in a different order.